Workflow to publish osdf-origin helm chart to Harbor (Software-5509)

jeff-takaki opened the pull request:
This workflow checks out the repo and runs through the steps that package the helm chart and pushes that helm chart to the Habor osg-htc project. The helm_package_path and helm_package_name variables work to grab the name of the packaged helm chart.

jeff-takaki committed:
Workflow to publish osdf-origin helm chart to Harbor (Software-5509)
This workflow checks out the repo and runs through the steps that package
the helm chart and pushes that helm chart to the Habor osg-htc project.
The helm_package_path and helm_package_name variables work to grab
the name of the packaged helm chart.

brianhlin reviewed (changes requested):
This is good for a first pass but there's still work to do in subsequent PRs:

1. We'll have multiple charts stored here so we'll need to only build + upload charts that have been changed in the future (see https://github.com/opensciencegrid/images/blob/main/.github/workflows/build-containers.yml for an example). This can go in a separate PR.
2. Where does `helm` get installed? I imagine that the `Publish Helm Chart` will fail when we merge this at this point!

jeff-takaki committed:
Add Step that installs Helm
Added step that gets the helm install script and executes the script.

jeff-takaki committed:
Update workflow to publish helm charts with changes
First attempt at trying to implement Brian's script to have the action only build and push helm charts
when there is a change within the directories.

jeff-takaki commented:
@brianhlin Ready for your eyes!

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+      - main
+
+jobs:
+  build-helm-list:
```
```suggestion
  build-chart-list:
```

Helm is to charts as Docker is to images. Please also update all the other references + vars

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+          # 2. For other branches (i.e., on someone's fork), diff main
+          #    vs the current commit
+          # 3. For PRs, diff the base ref vs the current commit
+          # 4. For everything else (e.g., dispatches), build all charts
```
Unlike our container images, we don't actually need to do regular mass rebuilds of charts so you can drop this step, i.e. drop the else statement from https://github.com/osg-htc/helm-charts/pull/4/files#diff-dce247c65e791dfbb2975e023ffc0d20d556f48be798bc6c462ba808ccbbc26fR51-R58

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+          
+          - name: Install Helm
+            run: |
+              curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
+              chmod 700 get_helm.sh
+              ./get_helm.sh
```
Oof, I don't love running a script that gets pulled off of the internet that then has access to various secrets in our environment. Is there a particular GHA that we can use or a specific binary that we can pull down instead?

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+              chmod 700 get_helm.sh
+              ./get_helm.sh
+
+          - name: Build Helm Chart
+            run: |
+              echo ${{secrets.OSG_HABOR_PASSWORD}} | helm registry login -u ${{secrets.OSG_HABOR_USER}} --password-stdin hub.opensciencegrid.org/osg-htc
+              helm_package_path=$(helm package supported/opensciencegrid/osdf-origin)
+              helm_package_name=$(basename "$helm_package_path")
+              helm push $helm_package_name oci://hub.opensciencegrid.org/osdf-origin
```
A few things:

1. There should be a step before this that attempts a build and this step should either only run on PRs if it duplicates the build that the push command is presumably doing, or run on both pushes/PRs if the push is smart enough to not rebuild if there's an existing build artifact
2. The chart to build is still hardcoded, meaning that this doesn't work when we add additional charts to `supported/osg-htc/`. For example, if we create `supported/osg-htc/jeffs-new-chart/` and merge that, then this will still /only/ build the `osdf-origin` chart
3. This job doesn't actually use the list of changed charts from the previous job meaning that even after you address (2), every merge into `main` will result in a rebuild and push of every chart regardless of it being changed or not. The images repo uses the list from the previous job to dynamically construct the matrix of images that need to be rebuilt and pushed: https://github.com/opensciencegrid/images/blob/main/.github/workflows/build-containers.yml#L107-L110

jeff-takaki replied:
@brianhlin For #1 do we want a step that's like a test build that's only runs if it's the same build as what is being pushed? Just a little confused on exactly what we want happening there.

brianhlin replied:
Basically we only want to push to hub.opensciencegrid.org when commits are pushed to the `main` branch on this fork, i.e. the repo is `osg-htc/helm-charts`. However, we still want useful CI (a `helm lint` and `helm package` are attempted) when 1) someone submits a PR against this repo and 2) when someone commits stuff on their own fork. 

Putting those together, we need to split up the package and push into two separate steps with appropriate `if:` expressions.

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+              echo ${{secrets.OSG_HABOR_PASSWORD}} | helm registry login -u ${{secrets.OSG_HABOR_USER}} --password-stdin hub.opensciencegrid.org/osg-htc
+              helm_package_path=$(helm package supported/opensciencegrid/osdf-origin)
+              helm_package_name=$(basename "$helm_package_path")
+              helm push $helm_package_name oci://hub.opensciencegrid.org/osdf-origin
```
```suggestion
              helm push $helm_package_name oci://hub.opensciencegrid.org/osg-htc/osdf-origin
```

I would think that we need to push to the `osg-htc` project? The hardcoded osdf-origin still needs to be fixed

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+
+          - name: Build Helm Chart
+            run: |
+              echo ${{secrets.OSG_HABOR_PASSWORD}} | helm registry login -u ${{secrets.OSG_HABOR_USER}} --password-stdin hub.opensciencegrid.org/osg-htc
```
```suggestion
              echo ${{secrets.OSG_HABOR_PASSWORD}} | helm registry login -u ${{secrets.OSG_HABOR_USER}} --password-stdin hub.opensciencegrid.org
```

This is possibly not right but it surprises me that you'd have to login to a specific "project" instead of the Harbor registry service as a whole.

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+
+      - id: helm-list
+        run: |
+          ORG_DIR=helm-charts
```
This isn't quite right as there's no `helm-charts` dir in the root of the repo https://github.com/osg-htc/helm-charts.

In the images repo, we were overloading the image list as A. the path to find the contents of each image, e.g. https://github.com/opensciencegrid/images/tree/main/opensciencegrid/sssd as well as B. the destination where to push the image, e.g. hub.opensciencegrid.org/**opensciencegrid/sssd**. In this repo, we have the complication of the leading "supported" vs "contrib" directories.

We still need all of (A) in the subsequent job to find the source of the build and we can still extract (B) from (A) to determine where the build needs to be pushed so I think the thing to do is populate the output of this job with the paths and then extract what you need in the subsequent job. In other words:

1. https://github.com/osg-htc/helm-charts/pull/4/files#diff-dce247c65e791dfbb2975e023ffc0d20d556f48be798bc6c462ba808ccbbc26fR38-R50 should result in full paths, e.g. `supported/osg-htc/osdf-origin/`
2. [This job](https://github.com/osg-htc/helm-charts/pull/4/files#diff-dce247c65e791dfbb2975e023ffc0d20d556f48be798bc6c462ba808ccbbc26fR85) should know how to get and use (A) and (B) from the paths that it's fed

jeff-takaki replied:
@brianhlin Is the idea here that the full path of the location of the helm chart is the same as where the chart is going to be pushed?

brianhlin replied:
Ah that's the complication with this repo when compared to the `images` repo. The full path of the location of the helm chart (e.g., `supported/opensciencegrid/osdf-origin/`) is not the same as the location of where we're going to be pushing the chart (e.g., `hub.opensciencegrid.org/opensciencegrid/osdf-origin`)

brianhlin commented on .github/workflows/publish-helm-chart.yml:
```diff
@@ -0,0 +1,90 @@
+name: Publish Helm Chart
+
+on:
+  push:
+    branches:
+      - main
```
We also want to run this on pull requests! In particular, we want to make sure that the helm checks out (`helm lint`) and builds `helm build`

jeff-takaki replied:
@brianhlin Do we want `helm lint` to run in the Build Helm Chart step? line 88? Is `helm build` the same as       `helm package`?

brianhlin replied:
Yeah, I think it'd be good to run `helm lint` in a step before we attempt the `helm package` (I thought it was `helm build` but I was wrong!)

brianhlin reviewed (commented):
Hey Jeff, nice attempt but there are some critical issues that need to be resolved before we can get this in. See comments for details.

jeff-takaki committed:
Rename job

Co-authored-by: Brian Lin <brianhlin@gmail.com>

jeff-takaki committed:
Fix project name for helm push

Co-authored-by: Brian Lin <brianhlin@gmail.com>

jeff-takaki committed:
Fix harbor login parameter

Co-authored-by: Brian Lin <brianhlin@gmail.com>

jeff-takaki committed:
Updated needs parameter in build push job to correct job reference

jeff-takaki committed:
Updated if component of build and push with correct job reference

jeff-takaki committed:
Fixed step for Helm install to use Azure GHA

jeff-takaki committed:
Removed else statement from chart list script.
Chart rebuilds unnecessary.

jeff-takaki committed:
Attempt to modify helmlist script to get full paths from supported and contrib dirs
I added what I think are the correct root dirs to variables that are both included in the script. I removed the cut line from the
script with the thinking that removing that line would yield the full paths for the chart directories instead of just the top directory.

jeff-takaki committed:
Modified build-and-push job to reference outputs from build-chart-list
This is an attempt to move away from hardcoded helm chart path in the helm package command. This should take in the
path generated by the build-chart-list script.

jeff-takaki committed:
Add pull request trigger

jeff-takaki committed:
Add Helm Chart Linter Step
Check before build if chart is well formed

jeff-takaki committed:
Split out Build and push steps
Added if statements so that linter and package run on pull requests and so that the chart
is pushed on commits to main branch

jeff-takaki committed:
Remove egrep and add cut line back to helm-list script

jeff-takaki committed:
Remove if statements for Build and Linter Steps

jeff-takaki committed:
Remove login command for Linter and Build steps
Add login command to Push step

jeff-takaki committed:
Clean up commands for Build step and re-work variables for Push step

jeff-takaki commented:
@brianhlin Might be a little better now?

brianhlin commented:
Ah dang, it looks like there are conflicts. Could you do the following?

1. Revert your rename of the dir
2. Rebase your branch on the `main` branch of this fork
3. Rename the dir again
4. Commit and push

jeff-takaki committed:
Fix merge conflict rename openscience grid dir to osg-htc

brianhlin reviewed (approved):
I think this looks pretty good! Let's see if it works!
